refactor(ast_tools): shorten generated code by avoiding `ref` in matches

overlookmotel opened the pull request:
Follow-on after #6404.

Shorten generated code for deriving `ESTree` by avoiding `ref` in matches.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6675?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6675** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6675?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6674** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6674?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6673** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6672** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6672?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

ottomated reviewed (approved):
If this works, it's fine with me—this existed to match serde's output as close as possible

overlookmotel committed:
refactor(ast_tools): shorten generated code by avoiding `ref` in matches (#6675)

Follow-on after #6404.

Shorten generated code for deriving `ESTree` by avoiding `ref` in matches.
